cpp/comms/std/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,407 @@
+  // Not supported by NCCL
+  ASSERT(false,
+         "ERROR: commSplit called but not yet supported in this comms "
+         "implementation.");
+}
+
+void cumlStdCommunicator_impl::barrier() const {
+  CUDA_CHECK(cudaMemsetAsync(sendbuff, 0, sizeof(int), stream));
+  CUDA_CHECK(cudaMemsetAsync(recvbuff, 0, sizeof(int), stream));
+
+  allreduce(sendbuff, recvbuff, 1, MLCommon::cumlCommunicator::INT,
+            MLCommon::cumlCommunicator::SUM, stream);
+
+  cudaStreamSynchronize(stream);
+}
+
+void cumlStdCommunicator_impl::get_request_id(request_t *req) const {
+#ifdef WITH_UCX
+
+  request_t req_id;
+
+  if (this->_free_requests.empty())
+    req_id = this->_next_request_id++;
+  else {
+    auto it = this->_free_requests.begin();
+    req_id = *it;
+    this->_free_requests.erase(it);
+  }
+  *req = req_id;
+#endif
+}
+
+void cumlStdCommunicator_impl::isend(const void *buf, int size, int dest,
+                                     int tag, request_t *request) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  get_request_id(request);
+
+  ucp_ep_h ep_ptr = (*_ucp_eps)[dest];
+
+  struct ucx_context *ucp_request =
+    ucp_isend(ep_ptr, buf, size, tag, getRank());
+
+  _requests_in_flight.insert(std::make_pair(*request, ucp_request));
+#endif
+
+  ASSERT(UCX_ENABLED, "cuML Comms not built with UCX support");
+}
+
+void cumlStdCommunicator_impl::irecv(void *buf, int size, int source, int tag,
+                                     request_t *request) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  get_request_id(request);
+
+  ucp_ep_h ep_ptr = (*_ucp_eps)[source];
+
+  struct ucx_context *ucp_request =
+    ucp_irecv(_ucp_worker, ep_ptr, buf, size, tag, source);
+
+  _requests_in_flight.insert(std::make_pair(*request, ucp_request));
+#endif
+
+  ASSERT(UCX_ENABLED, "cuML Comms not built with UCX support");
+}
+
+void cumlStdCommunicator_impl::waitall(int count,
+                                       request_t array_of_requests[]) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  std::vector<struct ucx_context *> requests;
+  requests.reserve(count);
+
+  for (int i = 0; i < count; ++i) {
+    auto req_it = _requests_in_flight.find(array_of_requests[i]);
+    ASSERT(_requests_in_flight.end() != req_it,
+           "ERROR: waitall on invalid request: %d", array_of_requests[i]);
+    requests.push_back(req_it->second);
+    _free_requests.insert(req_it->first);
+    _requests_in_flight.erase(req_it);
+  }
+
+  std::vector<struct ucx_context *> done_requests;
+  done_requests.reserve(count);
+
+  while (done_requests.size() < count) {
+    for (struct ucx_context *req : requests) {
+      ucp_worker_progress(_ucp_worker);
+
+      auto req_it = std::find(done_requests.begin(), done_requests.end(), req);
```

Comment:
Why not delete `req` from `requests` instead if it is completed and change the for loop in line 338 to a `while( requests.size() > 0 )` loop? I am bringing this up as I have concerns regarding the algorithmic complexity of the current solution and because I did not directly understood why the check was needed.

Reply:
Yep. My background as an enterprise Java developer expected concurrent modification problems while looping so I opted to use two vector's instead of removing directly from `requests`. I have fixed the code to loop using the iterator.